include/deal.II/matrix_free/matrix_free.h

```diff
@@ -1748,341 +2467,828 @@ reinit(const Mapping<dim>                                    &mapping,
+          const Utilities::MPI::Partitioner &part = get_partitioner(mf_component);
+          if (&part == matrix_free.get_dof_info(mf_component).vector_partitioner.get())
+            {
+              vec.compress_start(component_in_block_vector + channel_shift);
+              return;
+            }
+
+          if (part.n_ghost_indices()==0 && part.n_import_indices()==0)
+            return;
+
+          tmp_data[component_in_block_vector] = matrix_free.acquire_scratch_data_non_threadsafe();
+          tmp_data[component_in_block_vector]->resize_fast(part.n_import_indices());
+          AssertDimension(requests.size(), tmp_data.size());
+
+          part.import_from_ghosted_array_start
+          (dealii::VectorOperation::add,
+           component_in_block_vector+channel_shift,
+           ArrayView<Number>(vec.begin()+vec.get_partitioner()->local_size(),
+                             vec.get_partitioner()->n_ghost_indices()),
+           ArrayView<Number>(tmp_data[component_in_block_vector]->begin(),
+                             part.n_import_indices()),
+           this->requests[component_in_block_vector]);
+#endif
+        }
+    }
+
+    void compress_finish (const unsigned int component_in_block_vector,
+                          LinearAlgebra::distributed::Vector<Number> &vec)
+    {
+      (void)component_in_block_vector;
+      if (vector_face_access == dealii::MatrixFree<dim,Number>::DataAccessOnFaces::unspecified ||
+          vec.size() == 0)
+        vec.compress_finish(dealii::VectorOperation::add);
+      else
+        {
+#ifdef DEAL_II_WITH_MPI
+          AssertIndexRange(component_in_block_vector, tmp_data.size());
+          AssertDimension(requests.size(), tmp_data.size());
+
+          const unsigned int mf_component = find_vector_in_mf(vec);
+
+          const Utilities::MPI::Partitioner &part = get_partitioner(mf_component);
+          if (&part == matrix_free.get_dof_info(mf_component).vector_partitioner.get())
+            {
+              vec.compress_finish(dealii::VectorOperation::add);
+              return;
+            }
+
+          if (part.n_ghost_indices()==0 && part.n_import_indices()==0)
+            return;
+
+          part.import_from_ghosted_array_finish
+          (VectorOperation::add,
+           ArrayView<const Number>(tmp_data[component_in_block_vector]->begin(),
+                                   part.n_import_indices()),
+           ArrayView<Number>(vec.begin(), part.local_size()),
+           ArrayView<Number>(vec.begin()+vec.get_partitioner()->local_size(),
+                             vec.get_partitioner()->n_ghost_indices()),
+           this->requests[component_in_block_vector]);
+
+          matrix_free.release_scratch_data_non_threadsafe(tmp_data[component_in_block_vector]);
+          tmp_data[component_in_block_vector] = 0;
+#endif
+        }
+    }
+
+    void reset_ghost_values(const LinearAlgebra::distributed::Vector<Number> &vec) const
```

Comment:
This is kind of strange to have a function that resets values of a vector and yet takes the vector as const. That forces you to `const_cast` the vector for this function to be able to work... I would remove the const of the interface and, if necessary, do the `const_cast` before calling the function. This is just a weird interface.

Reply:
we discussed such things elsewhere on couple of occasions. Ghost part of la::d::Vector are really treated like mutables and the API of the vector itself has it in `zero_out_ghosts` and `update_ghost_values`.